[Rama Aryasuta Pangestu] iP

rama-pang opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

<hr>

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

rama-pang committed:
Added Level-1: Greet, Echo, Exit

rama-pang committed:
Add Level-2: Add, List

rama-pang committed:
Add Level-3: Mark as Done

rama-pang committed:
Add Level-4: ToDos, Events, Deadlines

rama-pang committed:
Add Automated Text UI Testing

rama-pang committed:
Add Level-5: Handle Errors

rama-pang committed:
Add Level-6: Delete

rama-pang committed:
Refactor code

rama-pang committed:
Add Level-7: Save

rama-pang committed:
Add Level-8: Dates and Times

rama-pang committed:
Merge branch 'branch-Level-7'

rama-pang committed:
Merge branch 'branch-Level-8'

rama-pang committed:
Refactor code

rama-pang committed:
Organize into packages

rama-pang committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

rama-pang committed:
Add A-Gradle

rama-pang committed:
Add JUnit Tests

rama-pang committed:
Remove emoji from Ui

rama-pang committed:
Add Level-9: Find

rama-pang committed:
Reformat code to comply with coding standard

rama-pang committed:
Add Javadocs

rama-pang committed:
Merge branch 'branch-Level-9'

rama-pang committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/task/Task.java

rama-pang committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java

rama-pang committed:
Reformat code

rama-pang committed:
Fix import order according to coding standard

rama-pang committed:
Add checkstyle

rama-pang committed:
Add GUI

rama-pang committed:
Merge branch 'branch-Level-10'

rama-pang committed:
Merge branch 'A-CheckStyle'

rama-pang committed:
Reformat fxml

pwjj2000 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+            ui.addError(e.getMessage());
+        }
+        if (isExit) {
+            System.exit(0);
```
Perhaps it would be better to further abstract this line out to a function? Rationale being new Java users might not be familiar with the properties of `System.exit(0)`

pwjj2000 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+import duke.ui.Ui;
+
+
+/**
+ * Duke, a Personal Assistant Chatbot that helps a person to keep track of various things.
+ *
+ * @author Rama Aryasuta Pangestu
+ */
```
Great JavaDocs documentation. :+1:
I have also noticed that you have done this consistently across all your classes and methods. Great job!

pwjj2000 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,123 @@
+ * @author Rama Aryasuta Pangestu
+ */
+public abstract class Task {
+    protected static final String ENCODING_SEPARATOR = " | ";
+    protected static final String DECODING_SEPARATOR = " \\| ";
```
Have you considered naming them `SEPARATOR_ENCODING` and `SEPARATOR_DECODING` instead?

pwjj2000 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,117 @@
+            break;
+        case "event":
+            try {
+                String description = Arrays.stream(args)
+                        .takeWhile(x -> !x.equals("/at")).skip(1)
+                        .reduce("", (x, y) -> x + " " + y).trim();
+                String[] timeArgument = Arrays.stream(args)
+                        .dropWhile(x -> !x.equals("/at")).toArray(String[]::new);
```
Perhaps this few lines can be shortened by abstracting them out into a helper function? Similar statements are found in other switch cases as well.

pwjj2000 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,117 @@
+                }
+                command = new AddCommand(
+                        new Event(description, false, LocalDate.parse(timeArgument[1])));
+            } catch (java.time.format.DateTimeParseException exception) {
```
Great catch of the exception here. Might it be better (and shorter) if `DateTimeParseException` is imported so that we do not need to include `java.time.format` at the front all the time?

pwjj2000 reviewed (commented):
Overall a great piece of work with no major issues!
Some minor changes can be done to improve readability of code.

rama-pang committed:
Refactor code: improve code quality

The `Duke` class directly calls `System.exit(0)` upon encountering
a `bye` user command.

New Java users might not be familiar with the properties
of `System.exit(0)`.

Let's abstract `System.exit(0)` to the `Launcher` class, as it is
the class responsible to initially start the application.

rama-pang committed:
Refactor code: improve code quality

In the `Parser` class, multiple command types (e.g., `find`, `todo`,
`deadline`, `event`) use `Arrays.stream` directly to parse each
argument.

The multiple uses of `Arrays.stream` causes code duplication.

Let's abstract argument-parsing to a helper function `parseArgument`.

rama-pang committed:
Add assertions

There are no assertions in code.

This might decrease code readability as there are some important
assumptions made that should hold in various points in the code.

Let's add assertions to the following classes:
- `Parser`
- `Deadline`
- `Event`
- `Task`
- `TaskList`
- `ToDo`

rama-pang committed:
Merge pull request #2 from rama-pang/branch-A-Assertions

Add assertions

rama-pang committed:
Merge branch 'master' into branch-A-CodeQuality

rama-pang committed:
Merge pull request #1 from rama-pang/branch-A-CodeQuality

branch-A-CodeQuality

rama-pang committed:
Create gradle.yml

rama-pang committed:
Add undo command support

There is no way to undo a command.

This might inconvenience the user if they made an unintentional command
or wants to revert their last command.

Let's add support for undo command. The undo command is executed if the
user enters `undo`.

The undo command will revert the last command which modifies the task
list. Note that the undo command can only undo a command which was made
since Duke is launched (i.e., there is no way to undo a command made in
a previous launch of Duke). Additionally, there is no way to undo
an undo command.

rama-pang committed:
Merge pull request #3 from rama-pang/branch-C-Undo

Add undo command support

rama-pang committed:
Refactor `Parser` class.

rama-pang committed:
Add welcome message from Duke

rama-pang committed:
Add credit for reused work

rama-pang committed:
Modify profile picture size in Duke

rama-pang committed:
Add Ui.png to docs/

rama-pang committed:
Add user guide

rama-pang committed:
Add user guide

rama-pang committed:
Add Ui.png to user guide
